### PR TITLE
Double deletion backend

### PR DESCRIPTION
**Pull request by MarcoLeipnitz**
This adds the double deletion method to the backend, which hides behind the usual DeletionMapping.

This also introduces some parts of our testing mechanisms.

closes #none yet (reference to #82)

---

### Implementation

- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified without unreasonable effort
- [ ] Is the code modular enough?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Single Responsibility Principle: Does any class serve only one purpose? If not, it should probably split up in multiple classes.
- [ ] Interface Segregation: If interface implemented, is it fully implemented?
- [ ] Dependency Injection: Are central object instantiation processes managed by a single context?
- [ ] If any functionality is missing, is it marked with a comment that starts with TODO:?
- [ ] Do you see any potential to improve the performance of the code significantly? 

### Logic Errors and Bugs

- [ ] Can you think of any use case in which the code does not behave as intended?
- [ ] Can you think of any inputs or external events that could break the code?
- [ ] Are edge cases covered by tests?

### Error Handling and Logging

- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information be added or removed?
- [ ] Are error messages user-friendly?

### Dependencies (internal and external)

- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Is the API.md up-to-date?
- [ ] Are the enumerations between frontend and backend synchronized?

### Security and Data Privacy

- [ ] Does the code introduce any security vulnerabilities?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is sensitive data like user data handled and stored securely?
- [ ] Is the right encryption used?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Were any secrets committed?

### Testing and Testability (later)

- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Do the existing tests reasonably cover the code change?
- [ ] Are there some test cases, input or edge cases that should be tested in addition?

### Usability and Accessibility

- [ ] Is the proposed solution (UI) accessible (e.g. colorblindness, ...)?
- [ ] Is the API/UI intuitive to use?
- [ ] Was the app tested on another device?

### Readability

- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by smaller methods/different names?
- [ ] Is the code located in the right file/folder?
- [ ] Do you think certain methods should be restructured to have a more intuitive control flow?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message better?
- [ ] Could some comments be removed by making the code itself more readable?

### CODE REVIEW

**Review by polodasi123 — DISMISSED** [review]

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/vladb/Appointment.java [hunk 1]
```diff
@@ -33,20 +42,27 @@ public class Appointment {
...
      */
     @Column(name = "notes", nullable = false)
     private String notes = "";
 
+    /**
+     * The id of a {@link User} that intents on deleting this appointment.
+     */
+    @Nullable
+    @Column(name = "deletingIntentionUser_id")
+    private Long deletingIntentionUserId;
```
[the remark this answers is not in the document]
Comment (JohaJung) [reply VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/vladb/Appointment.java]: Nein, aber das hier ist kein primitiver Typ (der wäre `long`, kleingeschrieben).

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/vladb/AppointmentSeries.java [hunk 2]
```diff
@@ -54,5 +58,11 @@ public class AppointmentSeries {
...
     private AppointmentCategory category;
 
-    // Getters and setters generated by Lombok @Data annotation.
+    /**
+     * List of appointments that this series references.
+     */
+    @JsonManagedReference
+    @OneToMany
+    @JsonIgnore
+    private List<Appointment> appointments;
```
Comment (JohaJung) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/vladb/AppointmentSeries.java]: The `@OneToMany` misses an argument `mappedBy = "series"`, since we have a manyToOne in the other direction already. This will avoid the redundant generated SQL table.

See https://docs.oracle.com/javaee/7/api/javax/persistence/OneToMany.html#mappedBy--

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/vladb/User.java [hunk 3]
```diff
@@ -40,6 +51,16 @@ public class User {
...
     private String password;
+
+    /**
+     * Represents all appointments that a user requested to delete.
+     */
+    @Nullable
+    @OneToMany
+    @JsonBackReference
+    @JsonIgnore
+    private List<Appointment> appointmentsWithDeletionIntention;
```
Comment (JohaJung) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/vladb/User.java]: Same as with appointments to appointmentSeries: add `mappedBy = "deletingIntentionUser"

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/vladb/Appointment.java [hunk 4]
```diff
@@ -67,4 +68,12 @@ public class Appointment {
...
     @OneToMany(mappedBy = "appointment", cascade = CascadeType.ALL)
     private List<ExperimentBooking> bookings = new ArrayList<>();
+
+    /**
+     * The id of a {@link User} that intents on deleting this appointment.
+     */
+    @Nullable
+    @ManyToOne // We avoid the @ManyToMany as we only need 2 approvals
+    @JsonManagedReference
+    private User deletingIntentionUser;
```
Comment (JohaJung) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/vladb/Appointment.java]: Doesn't this misses an `@JoinColumn(...` annotation specifying the column name? Same above for the `series` attribute.

See https://docs.oracle.com/javaee/7/api/javax/persistence/JoinColumn.html 

I'm confused that this did not cause errors at the series...

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/AppointmentController.java [hunk 5]
```diff
@@ -140,20 +167,50 @@ public ResponseEntity<Appointment[]> createAppointmentsMulti(
...
+        @PathVariable Long id
+    ) {
+        Appointment toDeleteAppointment = appointmentRepository.findById(id).orElseThrow(
             () -> new EntityNotFoundException(
-                "Appointment with ID " + id + " not found."));
+                "Appointment with ID " + id + " not found.")
+        );
 
-        if (!toBeDeletedAppointment.getBookings().isEmpty()) {
+        User deletingIntentUser = securityUtils.getCurrentUser();
```
Comment (JohaJung) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/AppointmentController.java]: Having `deletingIntentionUser` beneath `deletingIntentUser` is bad variable naming.

Not that I would care, but at Asims timeline PR you also complained today 😜

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/services/ExperimentBookingService.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            movedExperimentBookings.add(experimentBooking);
+        }
+
+        // Successfully moved experimentBookings
+        experimentBookingRepository.saveAll(movedExperimentBookings);
+    }
+
+    private void moveExperimentBookingsOnAppointmentWithLecture(
+        Appointment toBeDeletedAppointment,
+        List<ExperimentBooking> currentAppointmentExperimentBookings
```
Comment (JohaJung) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/services/ExperimentBookingService.java]: Am I too tired, or dose this function not respect the lecture at all? Did you overlooked a major logic flaw here? Mixed up both methods during a rewrite?

I.e., this method looks semantically equivalent to the `moveExperimentBookingsOnAppointmentWithoutLecture`, but coded differently (you used the Repository helper method instead of manually iterating the series).

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/security/VlaUserDetailsService.java [hunk 7]
```diff
@@ -32,6 +32,9 @@ public class VlaUserDetailsService implements UserDetailsService {
      * Interacts with the database and creates a {@link UserDetails}.
      * This object is used for checking the authentication credentials.
      * The "username" is the stringified id of the user.
+     *
+     * @param username The username of the user
+     * @return The {@link UserDetails} for Spring Boot Security
```
Comment (JohaJung) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/security/VlaUserDetailsService.java]: I would include the confusing line from above here:

„@return The UserDetails for Spring Boot Security, its "username" is the stringified ID of the user“
